subsys/bluetooth/audio/pacs.c

```diff
@@ -657,23 +699,151 @@ static sys_slist_t *pacs_get(enum bt_audio_dir dir)
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_PACS), \
+	BT_PAC_SNK(snk_read) \
+	BT_PACS_SNK_LOC(snk_loc_read) \
+	BT_PAC_SRC(src_read) \
+	BT_PACS_SRC_LOC(src_loc_read) \
+	BT_PAC_AVAILABLE_CONTEXT(available_contexts_read) \
+	BT_PAC_SUPPORTED_CONTEXT(supported_context_read) \
+}
+
+static struct bt_gatt_attr pacs_attrs[] = BT_PACS_SERVICE_DEFINITION();
+static struct bt_gatt_service pacs_svc = (struct bt_gatt_service)BT_GATT_SERVICE(pacs_attrs);
+
+
+
+
+#if defined(BT_PAC_SNK_NOTIFIABLE)
+#define PACS_SINK_PAC_CHAR_ATTR_COUNT 3  /* declaration + value + cccd */
+#else
+#define PACS_SINK_PAC_CHAR_ATTR_COUNT 2  /* declaration + value */
+#endif /* BT_PAC_SNK_NOTIFIABLE */
+
+#if defined(BT_PAC_SNK_LOC_NOTIFIABLE)
+#define PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT 3  /* declaration + value + cccd */
+#else
+#define PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT 2  /* declaration + value*/
+#endif /* BT_PAC_SNK_LOC_NOTIFIABLE */
+
+#if defined(BT_PAC_SRC_NOTIFIABLE)
+#define PACS_SOURCE_PAC_CHAR_ATTR_COUNT 3  /* declaration + value + cccd */
+#else
+#define PACS_SOURCE_PAC_CHAR_ATTR_COUNT 2  /* declaration + value */
+#endif /* BT_PAC_SRC_NOTIFIABLE */
+
+#if defined(BT_PAC_SRC_LOC_NOTIFIABLE)
+#define PACS_SOURCE_PAC_LOC_CHAR_ATTR_COUNT 3  /* declaration + value + cccd */
+#else
+#define PACS_SOURCE_PAC_LOC_CHAR_ATTR_COUNT 2  /* declaration + value*/
+#endif /* BT_PAC_SRC_LOC_NOTIFIABLE */
+
+static void configure_pacs_char(const struct bt_bap_pacs_register_param *param)
+{
+	size_t attrs_to_rem;
+	uint8_t first_to_rem;
+
+	/* Remove the Sink PAC and Location */
+	if (!param->snk_pac) {
+		first_to_rem = 0;
+		attrs_to_rem = PACS_SINK_PAC_CHAR_ATTR_COUNT + PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+	} else if (!param->snk_loc) {
+		first_to_rem = PACS_SINK_PAC_CHAR_ATTR_COUNT;
+		attrs_to_rem = PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+	} else {
+		first_to_rem = pacs_svc.attr_count;
+		attrs_to_rem = 0;
+	}
+
+	for (size_t i = first_to_rem + attrs_to_rem; i < pacs_svc.attr_count; i++) {
+		pacs_svc.attrs[i - attrs_to_rem] = pacs_svc.attrs[i];
+	}
+	pacs_svc.attr_count -= attrs_to_rem;
+
+	/* Set first_to_rem to the start of Source PAC Char, for cleaner offset calc */
+	first_to_rem = PACS_SINK_PAC_CHAR_ATTR_COUNT + PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+
+	/* Remove the Source PAC and Location */
+	if (!param->snk_pac) {
+		first_to_rem -= attrs_to_rem;
+		attrs_to_rem = PACS_SOURCE_PAC_CHAR_ATTR_COUNT +
+			       PACS_SOURCE_PAC_LOC_CHAR_ATTR_COUNT;
+	} else if (!param->snk_loc) {
+		first_to_rem = first_to_rem + PACS_SOURCE_PAC_CHAR_ATTR_COUNT - attrs_to_rem;
+		attrs_to_rem = PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+	} else {
+		return;
+	}
+
+	for (size_t i = first_to_rem; i < pacs_svc.attr_count; i++) {
+		pacs_svc.attrs[i - attrs_to_rem] = pacs_svc.attrs[i];
+	}
+	pacs_svc.attr_count -= attrs_to_rem;
+}
+
+int bt_pacs_register(const struct bt_bap_pacs_register_param *param)
+{
+	int err = 0;
+
+	if (atomic_test_bit(&pacs.flags, PACS_FLAG_REGISTERED)) {
```

Comment:
```suggestion
	if (atomic_test_and_set_bit(&pacs.flags, PACS_FLAG_REGISTERED)) {
```

`atomic_test_and_set_bit` must be used to be properly thread safe, otherwise 2 threads can pass this check before any of them sets the value. If any error then occurs, then the flag can be unset again